### DISCUSSION TITLE
Understanding a Verification Failure

### DISCUSSION THREAD

**Discussion by rcdickerson** [seahorn/seahorn#488] [Q&A] [answered]
Hello,

I am trying to understand why SeaHorn is failing to verify the following program:

```
#include "seahorn/seahorn.h"

extern int some_int();

int main(void) {
  int a1 = some_int();
  int a2 = some_int();
  assume(a1 == a2);

  int b1 = a1 * a1;
  int b2 = a2 * a2;

  sassert(b1 == b2);
 }
```

When I run `sea pf -m64 --cex=cex.ll program.c`, I see this output:

```
/usr/bin/clang-14 -c -emit-llvm -D__SEAHORN__ -fdeclspec -O1 -Xclang -disable-llvm-optzns -fgnu89-inline -m64 -I/home/usea/seahorn/include -o /tmp/sea-g3tgujq1/program.bc program.c
/home/usea/seahorn/bin/seapp -o /tmp/sea-g3tgujq1/program.pp.bc --simplifycfg-sink-common=false --strip-extern=false --promote-assumptions=false --kill-vaarg=true --ignore-def-verifier-fn=false --horn-keep-arith-overflow=false /tmp/sea-g3tgujq1/program.bc
/home/usea/seahorn/bin/seapp --simplifycfg-sink-common=false -o /tmp/sea-g3tgujq1/program.pp.ms.bc --horn-mixed-sem --ms-reduce-main /tmp/sea-g3tgujq1/program.pp.bc
/home/usea/seahorn/bin/seaopt -f --simplifycfg-sink-common=false -o /tmp/sea-g3tgujq1/program.pp.ms.o.bc -O3 --seaopt-enable-indvar=false --seaopt-enable-loop-idiom=false --unroll-threshold=150 --unroll-allow-partial=false --unroll-partial-threshold=0 --vectorize-slp=false /tmp/sea-g3tgujq1/program.pp.ms.bc
/home/usea/seahorn/bin/seahorn --keep-shadows=true --sea-dsa=ci --horn-solve -horn-cex-pass -horn-cex=cex.ll -horn-inter-proc -horn-sem-lvl=mem --horn-step=large /tmp/sea-g3tgujq1/program.pp.ms.o.bc
sat
```

The above output is from the llvm14 nightly docker image (`seahorn/seahorn-llvm14:nightly` on DockerHub). I have tried it on the llvm10 nightly build with a similar result.

 If I try to execute the counterexample, I see no output:
```
$ sea exe -m64 program.c cex.ll -o program
/usr/bin/clang-14 -c -emit-llvm -D__SEAHORN__ -fdeclspec -O1 -Xclang -disable-llvm-optzns -fgnu89-inline -m64 -I/home/usea/seahorn/include -o /tmp/tmp-tx28_rm4/program.bc program.c
/usr/bin/llvm-link-14 -o /tmp/sea-7612phld/merged.bc /tmp/tmp-tx28_rm4/program.bc cex.ll
/home/usea/seahorn/bin/seapp -o /tmp/sea-7612phld/merged.ext.bc --simplifycfg-sink-common=false --only-strip-extern=true --keep-lib-fn /tmp/sea-7612phld/merged.bc
/home/usea/seahorn/bin/seapp -o /tmp/sea-7612phld/merged.ext.int.bc --simplifycfg-sink-common=false --klee-internalize /tmp/sea-7612phld/merged.ext.bc
/home/usea/seahorn/bin/seapp -o /tmp/sea-7612phld/merged.ext.int.wmem.bc --wrap-mem /tmp/sea-7612phld/merged.ext.int.bc
/usr/bin/clang++-14 -m64 -o program /tmp/sea-7612phld/merged.ext.int.wmem.rmtf.bc /home/usea/seahorn/lib/libsea-rt.a
$ ./program
```

If I am reading the counterexample correctly (I'm not fluent in LLVM assembly), it looks like it is defining `some_int` to return zeroes:
```
...
@0 = private constant [2 x i32] zeroinitializer
@1 = private global i32 0
@__seahorn_cex_count = constant i32 2

define i32 @some_int(...) {
entry:
  %0 = load i32, i32* @1, align 4
  %1 = add i32 %0, 1
  store i32 %1, i32* @1, align 4
  %2 = call i32 @__seahorn_get_value_i32(i32 %0, i32* getelementptr inbounds ([2 x i32], [2 x i32]* @0, i32 0, i32 0), i32 2)
  ret i32 %2
}
...
```
If I manually change the values of `a1` and `a2` to zero then the program verifies. If I add the assignment `a1 = a2` after the `assume` in the original program, it also verifies. If I change `a1 * a1` and  `a2 * a2` to `a1` and `a2`, respectively, it verifies as well.

I'm new to SeaHorn, so I think it's likely I am making a simple mistake. I am quite stumped, though, and would appreciate any help understanding why this verification is failing. Thanks in advance!

**Comment by caballa** [comment, marked as answer]
This is a well know limitation of `pf`. This option runs the CHC solver using LIA to model integers. The advantage of `pf` is that it can infer automatically loop invariants and function summaries but it doesn't model integers using machine arithmetic among other limitations. For precise verification results, you can use the option `bpf` which will use the Bounded Model Checker instead. BMC will faithfully model integers using machine arithmetic. 

```
sea bpf --bmc=opsem test.c
```

should return unsat

**Reply by rcdickerson** [reply]
I see, thanks so much! One of my goals in using SeaHorn is to get automated loop invariant inference, is there a way to get invariant inference with programs whose VCs will fall outside of LIA?

**Reply by caballa** [reply]
Mostly LIA and arrays.

**Reply by caballa** [reply]
Although I think Spacer (the CHC solver) has also some support for bitvectors, right @agurfinkel ?

**Comment by agurfinkel** [comment]
There is an option `--horn-strictly-la=false` that allows non-linear arithmetic terms in generated CHCs. The default setting of the option is true, in which case, multiplication by non-constant is abstracted.

Support for non-linear arithmetic in Spacer (the CHC solver SeaHorn uses), and Z3 (the SMT solver Spacer uses) are best effort. Things should not crash. Some problems might be solved. However, there are no guarantees and the behaviour is not very predictable.

I know that people have used Spacer to solve CHC over non-linear arithmetic. The success depends on what exactly you are trying to do.

If you have a specific use case in mind, please let us know. Perhaps there is a solution we can work on. So far, the combination of the problem being really hard in general, and no specific use cases, have made non-linear arithmetic a low priority for invariant inference in SeaHorn.

As @caballa already mentioned, bit-vector arithmetic, linear and non-linear, are well supported in the BMC engine. There is some support for bit-vector arithmetic in CHC solver, but it is limited at best right now

**Reply by rcdickerson** [reply]
Thanks @agurfinkel, this flag seems like what I am looking for. 

My use case is evaluating a research project which aims to make loop invariant inference more tractable for product programs in the context of relational verification, so best effort is definitely acceptable. Incidentally, preliminaries of this work are appearing at the [E-Graphs workshop at FCRC](https://pldi23.sigplan.org/details/egraphs-2023-papers/10/KestRel-Relational-Verification-using-E-Graphs-for-Program-Alignment) later this month; if anyone is interested in more details please see the abstract and / or stop by the talk if you're at PLDI. :)

**Comment by agurfinkel** [comment]
Interesting. Two members of my group are attending the workshop and PLDI. They work on Spacer, and would be interested in learning more about your use case.

**Reply by rcdickerson** [reply]
More than happy to meet up to discuss! Email is in my Github profile, or otherwise I should be findable at the workshop.

**Reply by igcontreras** [reply]
@hgvk94 and me are going to present in the workshop :)

**Reply by rcdickerson** [reply]
Looks like we're even scheduled in consecutive timeslots!
